cpp/src/lists/contains.cu

```diff
@@ -44,384 +41,345 @@
- * @param stream CUDA stream used for device memory operations and kernel launches.
- */
-std::unique_ptr<column> contains_nulls(cudf::lists_column_view const& input_lists,
+std::unique_ptr<column> contains_nulls(lists_column_view const& lists,
                                        rmm::cuda_stream_view stream,
                                        rmm::mr::device_memory_resource* mr)
 {
-  auto const num_rows   = input_lists.size();
-  auto const d_lists    = column_device_view::create(input_lists.parent());
-  auto has_nulls_output = make_numeric_column(
-    data_type{type_id::BOOL8}, input_lists.size(), mask_state::UNALLOCATED, stream, mr);
-  auto const output_begin = has_nulls_output->mutable_view().begin<bool>();
-  thrust::tabulate(
-    rmm::exec_policy(stream),
-    output_begin,
-    output_begin + num_rows,
-    [lists = cudf::detail::lists_column_device_view{*d_lists}] __device__(auto list_idx) {
-      auto list       = list_device_view{lists, list_idx};
-      auto list_begin = thrust::make_counting_iterator(size_type{0});
-      return list.is_null() ||
-             thrust::any_of(thrust::seq, list_begin, list_begin + list.size(), [&list](auto i) {
-               return list.is_null(i);
-             });
-    });
-  auto const validity_begin = cudf::detail::make_counting_transform_iterator(
-    0, [lists = cudf::detail::lists_column_device_view{*d_lists}] __device__(auto list_idx) {
-      return not list_device_view{lists, list_idx}.is_null();
-    });
-  auto [null_mask, num_nulls] = cudf::detail::valid_if(
-    validity_begin, validity_begin + num_rows, thrust::identity<bool>{}, stream, mr);
-  has_nulls_output->set_null_mask(std::move(null_mask), num_nulls);
-  return has_nulls_output;
+  auto const lists_cv      = lists.parent();
+  auto output              = make_numeric_column(data_type{type_to_id<bool>()},
+                                    lists.size(),
+                                    copy_bitmask(lists_cv),
+                                    lists_cv.null_count(),
+                                    stream,
+                                    mr);
+  auto const out_begin     = output->mutable_view().template begin<bool>();
+  auto const lists_cdv_ptr = column_device_view::create(lists_cv, stream);
+
+  thrust::tabulate(rmm::exec_policy(stream),
+                   out_begin,
+                   out_begin + lists.size(),
+                   [lists = cudf::detail::lists_column_device_view{*lists_cdv_ptr}] __device__(
+                     auto const list_idx) {
+                     auto const list = list_device_view{lists, list_idx};
+                     return list.is_null() ||
+                            thrust::any_of(thrust::seq,
+                                           thrust::make_counting_iterator(0),
+                                           thrust::make_counting_iterator(list.size()),
+                                           [&list](auto const idx) { return list.is_null(idx); });
```

Comment:
Maybe you can check if the list has a null mask (is nullable) once, and then call `is_null_nocheck` in the algorithm. I'm not 100% sure if those APIs exist for `list_device_view` but I suspect they do.

Reply:
This is computing the output null mask for each row by checking if the input list row is null OR the list contains a null value. Another approach is to do a bitmask AND of a segmented nullmask reduction and the input (parent) null mask, then move-assign that to the output column's null mask. I have been meaning to write this into a utility function because the pattern is fairly common. I have a draft of this logic in #10451 but it needs to be simplified and put into a common location.

Reply:
Unfortunately that has not yet implemented in `list_device_view`. Maybe it is worth to have another follow up PR.

Reply:
Filed an FEA issue: https://github.com/rapidsai/cudf/issues/11072